### PR TITLE
Feature/processor additions

### PR DESCRIPTION
**Pull request by KasperOnPC**
Updated Processor Function with help from @thebonbon 
- Added ability to have multiple inputs
- Added ability to have multiple outputs
- Made it so items that cannot fit in the inventory are instead spawned at the player's feet.

### CODE REVIEW

File: src/Scripts/Game/Features/Processing/EL_ProcessAction.c [hunk 1]
```diff
@@ -1,81 +1,120 @@
...
+[BaseContainerProps()]
+class EL_ProcessingInput
+{
+    [Attribute(ResourceName.Empty, UIWidgets.ResourcePickerThumbnail, "Prefab to Input", "et")]
+	ResourceName m_InputPrefab;	
+	
+	[Attribute(defvalue: "1", uiwidget: UIWidgets.Auto, desc: "Input/s amount per process")]
+	int m_iInputAmount;
+	
+	ref array<IEntity> foundItems = new array<IEntity>();
```
Comment (thebonbon) [inline src/Scripts/Game/Features/Processing/EL_ProcessAction.c:10]:
```suggestion

```

File: src/Scripts/Game/Features/Processing/EL_ProcessAction.c [hunk 2]
```diff
@@ -1,81 +1,120 @@
...
+	[Attribute("", UIWidgets.Object, "List of inputs")]
+    ref array<ref EL_ProcessingInput> m_aProcessingInputs;
+	
+	[Attribute("", UIWidgets.Object, "List of outputs")]
+    ref array<ref EL_ProcessingOutput> m_aProcessingOutputs;
+
+	
+	
+	//------------------------------------------------------------------------------------------------
+	private array<IEntity> GetAllInputItems(InventoryStorageManagerComponent inventory, )
```
Comment (thebonbon) [inline src/Scripts/Game/Features/Processing/EL_ProcessAction.c:36]:
```suggestion
	private array<IEntity> GetAllInputItems(InventoryStorageManagerComponent inventory)
```

**Comment by Arkensor** [comment]
> Made it so items that cannot fit in the inventory are instead spawned at the player's feet.

Technically the best solution, the question would be if we should even allow the processing if the player can not store the output amount. I think we could do a pre-check with the player's current volume/weight and have some cached info on the volume and weight of the output prefabs to multiply by how much he will get.

Regardless of checks, I think an error message along the lines of "too full" could be a thing. So we would first try to add the output and if that fails to complete we only remove as much input as needed. We can update the system to do this later and go with spawn on feet for now, but during P1 I would like to get this sorted anyway.

File: src/Scripts/Game/Features/Processing/EL_ProcessAction.c [hunk 3]
```diff
@@ -1,81 +1,120 @@
...
+	{
+		array<IEntity> items = new array<IEntity>();
+		array<IEntity> foundItems = new array<IEntity>();
+		inventory.GetItems(items);
+
+		//Check if item in recipe
+		foreach (IEntity item : items)
+		{
+			foreach (EL_ProcessingInput m_aProcessingInput : m_aProcessingInputs)
```
Comment (thebonbon) [inline src/Scripts/Game/Features/Processing/EL_ProcessAction.c:45]:
```suggestion
			foreach (EL_ProcessingInput processingInput : m_aProcessingInputs)
```

Comment (thebonbon) [reply src/Scripts/Game/Features/Processing/EL_ProcessAction.c:45]: Not a member var

File: src/Scripts/Game/Features/Processing/EL_ProcessAction.c [hunk 4]
```diff
@@ -1,81 +1,120 @@
...
+	}
+
+	//------------------------------------------------------------------------------------------------
+	override void PerformAction(IEntity pOwnerEntity, IEntity pUserEntity)
+	{
+		InventoryStorageManagerComponent inventoryManager = InventoryStorageManagerComponent.Cast(pUserEntity.FindComponent(SCR_InventoryStorageManagerComponent));
+
+		array<IEntity> allInputItems = GetAllInputItems(inventoryManager);
+
+		foreach (EL_ProcessingInput m_aProcessingInput : m_aProcessingInputs)
```
Comment (thebonbon) [inline src/Scripts/Game/Features/Processing/EL_ProcessAction.c:64]:
```suggestion
		foreach (EL_ProcessingInput processingInput : m_aProcessingInputs) 
```

File: src/Scripts/Game/Features/Processing/EL_ProcessAction.c [hunk 5]
```diff
@@ -1,81 +1,120 @@
...
+			if (allInputItems.Count() < m_aProcessingInput.m_iInputAmount)
+				return;
+	
+			for (int i = 0; i < m_aProcessingInput.m_iInputAmount; i++) 
+			{
+				inventoryManager.TryDeleteItem(allInputItems[i]);
+			}
+		}
+		
+		bool CanSpawnToStorage = true;
```
Comment (thebonbon) [inline src/Scripts/Game/Features/Processing/EL_ProcessAction.c:75]:
```suggestion
		bool bCanSpawnToStorage = true;
```

File: src/Scripts/Game/Features/Processing/EL_ProcessAction.c [hunk 6]
```diff
@@ -1,81 +1,120 @@
...
+		
+		foreach (EL_ProcessingOutput m_aProcessingOutput : m_aProcessingOutputs)
```
Comment (thebonbon) [inline src/Scripts/Game/Features/Processing/EL_ProcessAction.c:77]:
```suggestion
		foreach (EL_ProcessingOutput processingOutput : m_aProcessingOutputs) 
```

File: src/Scripts/Game/Features/Processing/EL_ProcessAction.c [hunk 7]
```diff
@@ -1,81 +1,120 @@
...
+		bool CanSpawnToStorage = true;
+		
+		foreach (EL_ProcessingOutput m_aProcessingOutput : m_aProcessingOutputs) 
+		{
+			for (int i = 0; i < m_aProcessingOutput.m_iOutputAmount; i++) 
+			{
+				CanSpawnToStorage = inventoryManager.TrySpawnPrefabToStorage(m_aProcessingOutput.m_OutputPrefab);
+				if (!CanSpawnToStorage)
+				{
+					EntitySpawnParams spawnParams = new EntitySpawnParams();
```
Comment (thebonbon) [inline src/Scripts/Game/Features/Processing/EL_ProcessAction.c:84]: Use `EL_Utils.SpawnEntityPrefab`.. here

**Comment by KasperOnPC** [comment]
Yeah, we came to use spawn at feet because the only real current solution was the one where you have to spawn the prefab in the "debug world" to get its entity to get its inventory size. With multiple outputs, this was requiring the spawning of an array of items in the "debug world" which seemed super dodgy.

I agree a dialog box should at least be added that says inventory full when you process and the item is spawned on the ground. I'm happy to be assigned that as an issue for later but would love if this can be merged as-is and that can be added once I / someone else has time to learn about the dialog box prompts and implement it here.

File: src/Scripts/Game/Features/Processing/EL_ProcessAction.c [hunk 8]
```diff
@@ -1,81 +1,120 @@
...
+		}
+	}
+
+	//------------------------------------------------------------------------------------------------
+	override bool CanBePerformedScript(IEntity user)
+ 	{
+		//Check player inventory
+		InventoryStorageManagerComponent inventoryManager = InventoryStorageManagerComponent.Cast(user.FindComponent(SCR_InventoryStorageManagerComponent));
+		
+		array<string> CannotPerformArray = {};
```
Comment (thebonbon) [inline src/Scripts/Game/Features/Processing/EL_ProcessAction.c:101]: What is this used for?

File: src/Scripts/Game/Features/Processing/EL_ProcessAction.c [hunk 9]
```diff
@@ -1,81 +1,120 @@
...
+		CannotPerformArray.Insert("Hello");
+		
+		bool CanPerform = true;
```
Comment (thebonbon) [inline src/Scripts/Game/Features/Processing/EL_ProcessAction.c:104]:
```suggestion
		bool bCanPerform = true;
```

File: src/Scripts/Game/Features/Processing/EL_ProcessAction.c [hunk 10]
```diff
@@ -1,81 +1,120 @@
...
+		
+		foreach (EL_ProcessingInput m_aProcessingInput : m_aProcessingInputs)
```
Comment (thebonbon) [inline src/Scripts/Game/Features/Processing/EL_ProcessAction.c:106]:
```suggestion
		foreach (EL_ProcessingInput processingInput : m_aProcessingInputs)
```

**Review by thebonbon — CHANGES_REQUESTED** [review]

**Review by thebonbon — APPROVED** [review]

**Comment by KasperOnPC** [comment]
Updated to meet the script conventions @Arkensor, anything else required?

**Review by Arkensor — CHANGES_REQUESTED** [review]
Minor formatting changes

**Review by Arkensor — APPROVED** [review]

**Pull request merged**